src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs

```diff
@@ -0,0 +1,313 @@
+        /// <remarks>
+        /// The default implementation of this method looks for a <c>public static IWebHostBuilder CreateDefaultBuilder(string[] args)</c>
+        /// method defined on the entry point of the assembly of <typeparamref name="TEntryPoint" /> and invokes it passing an empty string
+        /// array as arguments. In case this method can't be found,
+        /// </remarks>
+        /// </summary>
+        /// <returns>A <see cref="IWebHostBuilder"/> instance.</returns>
+        protected virtual IWebHostBuilder CreateWebHostBuilder() =>
+            WebHostBuilderFactory.CreateFromTypesAssemblyEntryPoint<TEntryPoint>(Array.Empty<string>()) ??
+            throw new InvalidOperationException($"No method 'public static {nameof(IWebHostBuilder)} CreateWebHostBuilder(string[] args)' " +
+                $"found on '{typeof(TEntryPoint).Assembly.EntryPoint.DeclaringType.FullName}'. Alternatively, " +
+                $"{typeof(WebApplicationFactory<TEntryPoint>).Name} can be extended and 'protected virtual {nameof(IWebHostBuilder)} " +
+                $"{nameof(CreateWebHostBuilder)}()' can be overriden to provide your own {nameof(IWebHostBuilder)} instance.");
+
+        /// <summary>
+        /// Creates the <see cref="TestServer"/> with the bootstrapped application in <paramref name="builder"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IWebHostBuilder"/> used to
+        /// create the server.</param>
+        /// <returns>The <see cref="TestServer"/> with the bootstrapped application.</returns>
+        protected virtual TestServer CreateServer(IWebHostBuilder builder) => new TestServer(builder);
+
+        /// <summary>
+        /// Gives a fixture an opportunity to configure the application before it gets built.
+        /// </summary>
+        /// <param name="builder">The <see cref="IWebHostBuilder"/> for the application.</param>
+        protected virtual void ConfigureWebHost(IWebHostBuilder builder)
+        {
+        }
+
+        private void EnsureBuilder()
+        {
+            if (_builder != null)
+            {
+                return;
+            }
+
+            _builder = CreateWebHostBuilder();
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="HttpClient"/> that automatically follows
+        /// redirects and handles cookies.
+        /// </summary>
+        /// <returns></returns>
+        public HttpClient CreateStandardClient() =>
+            CreateClient(new RedirectHandler(), new CookieContainerHandler());
+
+        /// <summary>
+        /// Creates a new instance of an <see cref="HttpClient"/> that can be used to
+        /// send <see cref="HttpRequestMessage"/> to the server.
+        /// </summary>
+        /// <returns>The <see cref="HttpClient"/></returns>
+        public HttpClient CreateClient(params DelegatingHandler[] handlers)
+        {
+            return CreateClient(new Uri("http://localhost"), handlers);
```

Comment:
I don't remember if we talked about this before, but how should we think about `http` vs. `https`?

Reply:
By default we just do http, we can't generally assume that the app is set up for using https and the same effect can be achieved by calling the overload that allows you to specify a base address. As an alternative we could expose the base address in a settable property, but I would at this point wait for feedback on it. (which I don't expect)